src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCaseEquivalences.cs

```diff
@@ -0,0 +1,167 @@
+        };
+
+        /// <summary>
+        /// Performs a fast lookup which determines if a character is involved in case conversion, as well as
+        /// returns the characters that should be considered equivalent in case it does participate in case conversion.
+        /// This method, in contrast to <see cref="TryFindCaseEquivalencesForChar(char, out ReadOnlySpan{char})"/> does take in
+        /// culture and will also factor in the current culture in order to handle the special cases which are different between cultures.
+        /// </summary>
+        /// <param name="c">The character being analyzed</param>
+        /// <param name="culture">The <see cref="CultureInfo"/> to be used to determine the equivalences.</param>
+        /// <param name="equivalences">If <paramref name="c"/> is involved in case conversion, then equivalences will contain the
+        /// span of character which should be considered equal to <paramref name="c"/> in a case-insensitive comparison.</param>
+        /// <returns><see langword="true"/> if <paramref name="c"/> is involved in case conversion; otherwise, <see langword="false"/></returns>
+        public static bool TryFindCaseEquivalencesForCharWithIBehavior(char c, CultureInfo culture, out ReadOnlySpan<char> equivalences)
+        {
+            if ((c | 0x20) == 'i' || (c | 0x01) == '\u0131')
+            {
+                RegexCaseBehavior mappingBehavior = GetRegexBehavior(culture);
+                equivalences = c switch
+                {
+                    // Invariant mappings
+                    'i' or 'I' when mappingBehavior is RegexCaseBehavior.Invariant => s_specialCasingSetBehaviors.AsSpan(0, 2), // 'I' and 'i'
+
+                    // Non-Turkish mappings
+                    'i' or 'I' or '\u0130' when mappingBehavior is RegexCaseBehavior.NonTurkish => s_specialCasingSetBehaviors.AsSpan(0, 3), // 'I', 'i', and '\u0130'
+
+                    // Turkish mappings
+                    'I' or '\u0131' when mappingBehavior is RegexCaseBehavior.Turkish => s_specialCasingSetBehaviors.AsSpan(3, 2), // 'I' and '\u0131'
+                    'i' or '\u0130' when mappingBehavior is RegexCaseBehavior.Turkish => s_specialCasingSetBehaviors.AsSpan(1, 2), // 'i' and '\u0130'
+
+                    // Default
+                    _ => null
+                };
+                return equivalences != null;
+            }
+            else
+            {
+                return TryFindCaseEquivalencesForChar(c, out equivalences);
+            }
+        }
+
+        /// <summary>
+        /// Returns which <see cref="RegexCaseBehavior"/> should be used based on the passed in <paramref name="culture"/>.
+        /// </summary>
+        /// <param name="culture">The <see cref="CultureInfo"/> to be used to determine the behavior.</param>
+        /// <returns>The <see cref="RegexCaseBehavior"/> that should be used.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Why AggressiveInlining?

Reply:
Given this is a hot path, I was following precedence of what we do today for ToLower/ToUpper:

https://github.com/dotnet/runtime/blob/c0db07b3f3f093b5f4cd27e1f5e8aa54adad049d/src/libraries/System.Private.CoreLib/src/System/Globalization/CharUnicodeInfo.cs#L285-L297

That said, I didn't validate if the method actually needs it or if it gets inlined automatically even without the annotation.